### PR TITLE
Design doc: Deployment procedure actions

### PR DESCRIPTION
**Pull request by rohantmp**
**Describe what this PR does**
Starts a design doc that contains actions to implement.

### CODE REVIEW

**Comment by rohantmp** [comment]
@JohnStrunk

File: docs/Developers/Design/deploy_procedure_design.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+## GlusterNode actions
+
+`GlusterNode` CRs can be created manually or by the `GlusterCluster` Controller according to a `template`. The node is associated with the cluster using the `gluster.org/cluster-name` label that they will both share. `GlusterCluster`s that consume local storage via `hostPath` require their `GlusterNode`s to be created manually and have `nodeAffinity` set in a way that it will only be scheduled on that node.
+
+### statefulSetReconciled
+prereqs:
+ -  [managedDevicesReconciled](#managedDevicesReconciled)
+### managedDevicesReconciled
+prereqs:
+ -  [glusterClusterServicesReconciled](#glusterClusterServicesReconciled)
```
Comment (JohnStrunk) [inline docs/Developers/Design/deploy_procedure_design.md]: Is there a circular dependency here?
`managedDevices -> services -> nodes ?->? managedDevices`

**Review by JohnStrunk — CHANGES_REQUESTED** [review]
This needs a rebase as it contains already merged code, i believe.

General comment:
I think there are 2 objectives for this doc:
1. Describe the DAG of actions/prereqs so that someone can get an overview of the flow.
2. Describe what the actions are designed to do.

For describing the DAG, consider using `dot`. It will make any cycles obvious, and give a good visual view. Example: https://github.com/gluster/anthill/blob/master/docs/Developers/Design/crd_hierarchy.dot

Eventually, I'd like to see a tool that pulls the code comments for the actions/prereqs to generate a readable document explaining the purpose (if could also generate the dot file too) :smile: 
For now, this manual doc is probably fine. I just worry about making sure it is consistent w/ the code comments (and both are up-to-date).

**Comment by rohantmp** [comment]
Rebased! My bad( I accidentally branched off of the branch where I was experimentally squashing commits for my old PR).

 +1 on the objectives and DAG. I was thinking the same thing about dot after learning about it from the same doc you mentioned.

File: docs/Developers/Design/deploy_procedure_design.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
+This document describes the actions that have to be executed successfully for both of anthill's CRDs to be consider reconciled.  They are implemented as `reconciler.Action` objects, and are enumerated in a `reconciler.Procedure` object. Procedure level actions are ones that modify state and should have a corresponding entry populated in `.Status.ReconcileActions map[string]reconciler.Result`. Top level actions are executed in an arbitrary order so they must define any prerequisite actions explicitly. An action may be a top-level action and still defined as a prerequisite and the caching implementation will ensure that it is executed only the necessary amount of times(once?).
```
[the remark this answers is not in the document]
Comment (rohantmp) [reply docs/Developers/Design/deploy_procedure_design.md]: +1

File: docs/Developers/Design/deploy_procedure_design.md (hunk 1 above)
Comment (rohantmp) [reply docs/Developers/Design/deploy_procedure_design.md]: Removing it from the node actions.

**Review by JohnStrunk — APPROVED** [review]

**Pull request merged**